src/Bicep.Core/Parsing/BaseParser.cs

```diff
@@ -0,0 +1,1185 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+
+namespace Bicep.Core.Parsing
+{
+    public abstract class BaseParser
+    {
+        protected readonly TokenReader reader;
+
+        protected readonly ImmutableArray<IDiagnostic> lexerDiagnostics;
+
+        protected BaseParser(string text)
+        {
+            // treating the lexer as an implementation detail of the parser
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+            var lexer = new Lexer(new SlidingTextWindow(text), diagnosticWriter);
+            lexer.Lex();
+
+            this.lexerDiagnostics = diagnosticWriter.GetDiagnostics().ToImmutableArray();
+
+            this.reader = new TokenReader(lexer.GetTokens());
+        }
+
+        public abstract ProgramSyntax Program();
+
+        protected abstract SyntaxBase Declaration();
```

Comment:
>protected [](http://example.com/codeflow?start=8&length=9)

This used to be `public`, but doesn't need to be. Made it abstract because  it's referenced in `ObjectProperty()`.